Make SharedStorageMountDirValidator accept /scratch and improve error message

hanwen-pcluste opened the pull request:
### Description of changes
This is an amendament of https://github.com/aws/aws-parallelcluster/pull/3981.
See commit descriptions for details

Signed-off-by: Hanwen <hanwenli@amazon.com>

### Tests
The unit test has been improved to test the logic. 
`test_ebs_multiple` is improved to test `/scratch` used by a shared EBS on instances without instance store

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

hanwen-cluster committed:
Make SharedStorageMountDirValidator accept /scratch

This is an amendament of https://github.com/aws/aws-parallelcluster/pull/3981.

Specifically, the previous PR always restricts the usage of `/scratch` on [shared storage](https://docs.aws.amazon.com/parallelcluster/latest/ug/SharedStorage-v3.html) to prevent overwriting ephemeral drive mount directory. However, only a subset of instance types has [ephemeral drive (instance store)](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/InstanceStorage.html). Therefore, we propose to only print out an error when `/scratch` is used by shared storage AND some of the head node and compute nodes support ephemeral drive and use `/scratch` as the mount directory.

Signed-off-by: Hanwen <hanwenli@amazon.com>

gmarciani commented on cli/src/pcluster/aws/aws_resources.py:
```diff
@@ -205,6 +205,10 @@ def vcpus_count(self) -> int:
 
+    def instance_storage_supported(self) -> bool:
+        """Indicate whether instance storage is supported."""
+        return self.instance_type_data.get("InstanceStorageSupported")
```
Is the field `InstanceStorageSupported ` always present in `instance_type_data`?
If we are not sure about it, we should consider to return false by default.

hanwen-pcluste replied:
It comes from the doc of boto3 https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.describe_instance_types. So I believe this is strong enough to assume the field always exist

gmarciani commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1331,19 +1332,24 @@ def shared_mount_dir_list(self):
+                if self.head_node.local_storage.ephemeral_volume
+                else DEFAULT_EPHEMERAL_DIR
+            }
+            if self.head_node.instance_type_info.instance_storage_supported()
```
Correct me if I'm wrong.
This change makes me think that we are fixing a bug that we had regardless of the main goal of this PR.
In particular, we were assuming to always have a local mount dir set even if the instance was not supporting instance storage. Is this right?

hanwen-pcluste replied:
You can call it a bug.

gmarciani commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1905,6 +1918,13 @@ def instance_type_list(self):
         """Return the list of instance types associated to the Queue."""
         return [compute_resource.instance_type for compute_resource in self.compute_resources]
 
+    def has_instance_storage(self):
```
We are duplicating the same `has_instance_storage` function in both `SchedulerPluginQueue` and `SlurmQueue`. So, what about adding the function in `_CommonQueue`, instead?

gmarciani replied:
This duplication is actually a signal of a defect in the aforementioned class design.
By the way, this would required a refactoring that is out of scope for this PR.
So, let's proceed with the proposed implementation, but let's be sure to track the need for improvement.

gmarciani commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -589,6 +588,13 @@ def _validate(self, mount_dir_list):
+                self._add_failure(
+                    f"Ephemeral drive mount directory `{local_mount_dir}` is overwritten by shared storage "
+                    f"mount directory. Please consider changing either directory to avoid the overwrite.",
+                    FailureLevel.WARNING,
```
**Product decision alert**
I think that this kind of failure should be an error rather than a warning. Overwriting a local mount point should block the cluster creation in the same way a shared mount point overwrites another shared mount point.

Pending product decision here.

gmarciani replied:
Product decision is to be more conservative and keep the warning level.

gmarciani commented on cli/tests/pcluster/validators/test_all_validators.py:
```diff
@@ -224,14 +224,14 @@ def test_validators_are_called_with_correct_argument(test_datadir, mocker):
     )
     key_pair_validator.assert_has_calls([call(key_name="ec2-key-name")])
-    instance_type_validator.assert_has_calls([call(instance_type="t2.micro")])
+    instance_type_validator.assert_has_calls([call(instance_type="c5d.xlarge")])
```
It's not clear to me why you are removing the t2.micro case.
I was expecting to add tests/assertions not to replace existing ones?
May you please clarify?

hanwen-pcluste replied:
This is to use a instance type that has an instance store

gmarciani replied:
Got it: so we are replacing t2.micro with c5d.xlarge because the paths triggered by the latter is a superset of the paths triggered by the former

gmarciani commented on cli/tests/pcluster/validators/test_cluster_validators.py:
```diff
@@ -827,32 +827,43 @@ def test_fsx_architecture_os_validator(architecture, os, expected_message):
             "Mount directories dir2, dir1 cannot be specified for multiple file systems",
         ),
+        (
+            ["dir1", "dir2", "/scratch"],
```
Since `/scratch` is a notable mount point, I think it's worth it to have a test case verifying that when shared_mount_dir_list contains `/shared` and local mount dir does not, the validator succeeds

hanwen-pcluste replied:
I included the test case in integration test.

gmarciani replied:
I think we should add it in unit tests because:
1. validator logic is expected to be covered by unit tests 
2. this is a notable test case for the validator
3. if something can be tested both with unit tests and integ test, better to have in unit test at forst because they are faster and cheaper

hanwen-pcluste replied:
Done

hanwen-cluster committed:
Improve error message of DuplicateMountDirValidator

To improve the message, more information needs to be passed to the validator. Therefore, this commit refactors relevant code

New message:
duplicate between shared storage and local storage:
```
{
      "level": "ERROR",
      "type": "DuplicateMountDirValidator",
      "message": "The mount directory `/scratch` used for shared storage ['my-ebs'] clashes with the one used for ephemeral store volumes of the instances ['c5ad.xlarge', 'c5d.xlarge']. Please change the mount directory configuration of either the shared storage or the ephemeral volume of the impacted nodes."
}
```
duplicate between shared directories:
```
{
      "level": "ERROR",
      "type": "DuplicateMountDirValidator",
      "message": "The mount directory `/shared` is used for multiple shared storage: ['my-ebs', 'my-fsx']. Shared storage mount directories should be unique. Please change the mount directory configuration of the shared storage."
}
```

Old message:
duplicate between shared storage and local storage:
```
{
      "level": "ERROR",
      "type": "DuplicateMountDirValidator",
      "message": "Ephemeral drive mount directory `/scratch` is overwritten by shared storage mount directory. Please consider changing either directory to avoid the overwrite."
}
```
duplicate between shared directories:
```
{
      "level": "ERROR",
      "type": "DuplicateMountDirValidator",
      "message": "Mount directory /shared cannot be specified for multiple file systems"
}
```

Signed-off-by: Hanwen <hanwenli@amazon.com>

gmarciani commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1320,32 +1321,37 @@ def partition(self):
-            else DEFAULT_EPHEMERAL_DIR
-        }
+    def local_mount_dir_instance_types_dict(self):
+        """Retrieve a dictionary of local mount dirs and corresponding instance types."""
```
Why do we need a mapping between local mount dir and instance types?
Isn't it enough to have a list of local mount dir?

hanwen-pcluste replied:
Because we want the error message to include instance type information

gmarciani replied:
Got it, that's fine

gmarciani commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1320,32 +1321,37 @@ def partition(self):
+            mount_dir_instance_types_dict[
+                self.head_node.local_storage.ephemeral_volume.mount_dir
+                if self.head_node.local_storage.ephemeral_volume
+                else DEFAULT_EPHEMERAL_DIR
```
Minor: this logic is implemented the same both for head node and compute nodes. We should consider to put it in common and reduce code duplication.
May you please keep track of this refactoring in our backlog?

gmarciani commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -581,16 +581,28 @@ class DuplicateMountDirValidator(Validator):
+            if shared_storage_names:
+                self._add_failure(
+                    f"The mount directory `{local_mount_dir}` used for shared storage {shared_storage_names} "
+                    f"clashes with the one used for ephemeral store volumes of the instances {list(instance_types)}. "
```
We refer to "ephemeral store volume", but AWS docs typically refer to this storage as "instance store volume".
Furthermore, we refer to that one time as "ephemeral store volume" and another time as "ephemeral volume", let's use unique naming.

hanwen-pcluste replied:
Done. Thanks

gmarciani commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -581,16 +581,28 @@ class DuplicateMountDirValidator(Validator):
-                ),
-                FailureLevel.ERROR,
-            )
+    def _validate(self, shared_storage_name_mount_dir_tuple_list, local_mount_dir_instance_types_dict):
```
In this validation we do a two steps verification:
1. first verify that there is no clash among shared storages
2. then, verify that there is no clash between shared storages and local storages.
Did you do it this way because the goal was to provide distinct messages for shared v/s shared and shared v/s local?

hanwen-pcluste replied:
Yes, because of the different messages

gmarciani commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1905,6 +1920,15 @@ def instance_type_list(self):
+        """Return a set of instance types in the queue that have instance store."""
+        result = set()
+        for compute_resource in self.compute_resources:
+            if compute_resource.instance_type_info.instance_storage_supported():
```
Does this introduce the risk of having a lot of calls to EC2 to retrieve such information?
In case it does, we should cache the info about instance storage support.

hanwen-pcluste replied:
There is cache for instance_type_info. So no duplicate EC2 is required

gmarciani replied:
Great, thanks
